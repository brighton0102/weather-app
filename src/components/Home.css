@import url('https://fonts.googleapis.com/css2?family=Gothic+A1:wght@100;200;300;400;500;600;700;800;900&family=Lato:wght@100;300;400;700&family=Montserrat:wght@100;200;300;400;500;600&family=Permanent+Marker&family=Poppins:ital,wght@0,100;0,300;0,400;0,500;0,600;1,100&family=Space+Grotesk:wght@300;400;500;600;700&family=Titan+One&display=swap');

* {
  margin: 0;
  padding: 0;
  font-family: 'Poppins', sans-serif;
  color: white;
}

body {
  background-image: url('../features/maroon.jpg');
  background-repeat: no-repeat;
  background-size: cover;
  background-position: center;
  background-attachment: fixed;
}

input[type=text] {
  width: 75%;
  padding: 10px 15px;
  margin: 20px 20px;
  box-sizing: border-box;
  font-size: 18px;
}

.title {
  padding-top: 20px;
  font-size: 20px;
}

.container {
  display: flex;
  flex-direction: column;
  align-items: center;
  text-align: center;
}

.angle-arrow {
  color: white;
  font-size: 20px;
  position: relative;
  left: 45%;
}

.cards {
  list-style: none;
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
}

.card {
  padding: 8px;
  width: calc(50% - 24px);
  box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
  background-color: rgba(202, 69, 111, 1);
  display: inline-block;
  vertical-align: top;
}

.card p {
  padding: 1px 0;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

.details-link {
  display: block;
  text-align: right;
  margin-top: 10px;
  color: #007bff;
  text-decoration: none;
  font-weight: bold;
}

.details-link:hover {
  text-decoration: underline;
}

.loading-spinner {
  border: 4px solid rgba(0, 0, 0, 0.3);
  border-top: 4px solid black;
  border-radius: 50%;
  width: 24px;
  height: 24px;
  animation: spin 1s linear infinite;
  position: relative;
  left: 50%;
}

@keyframes spin {
  0% { transform: rotate(0deg); }
  100% { transform: rotate(360deg); }
}

@media screen and (min-width: 768px) {
  .card {
    width: 320px;
  }

  input[type=text] {
    width: 25%;
    padding: 12px 20px;
    margin: 20px 20px;
    box-sizing: border-box;
    font-size: 20px;
  }
}
