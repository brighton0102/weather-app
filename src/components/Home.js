import React, { useState, useEffect } from 'react';
import { Link } from 'react-router-dom';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faArrowCircleRight } from '@fortawesome/free-solid-svg-icons';
import fetchWeatherByCityName from '../services/api';
import './Home.css';

function Home() {
  const [cities, setCities] = useState(['New York City', 'Tokyo', 'London', 'Paris', 'Beijing', 'Los Angeles', 'Moscow', 'Istanbul',
    'Dubai', 'Singapore', 'Sydney', 'Rio de Janeiro', 'Mumbai', 'Rome', 'Toronto', 'Buenos Aires', 'Cairo', 'Hong Kong', 'Amsterdam', 'Seoul', 'Barcelona', 'San Francisco', 'Chicago', 'Bangkok', 'Berlin', 'Kuala Lumpur', 'Dar es salaam', 'Vancouver', 'Mombasa', 'Athens', 'Cape Town', 'Prague']);
  const [filteredCities, setFilteredCities] = useState(cities);
  const [selectedCity, setSelectedCity] = useState('');
  const [weatherData, setWeatherData] = useState({}); // Initialize as an empty object

  useEffect(() => {
    // Load weather data for all cities in the array
    cities.forEach(async (city) => {
      try {
        const data = await fetchWeatherByCityName(city);
        setWeatherData((prevData) => ({ ...prevData, [city]: data }));
      } catch (error) {
        console.error(error);
      }
    });
    const cards = document.querySelectorAll('.card');

    cards.forEach((card, index) => {
      if (index % 4 === 0 || index % 4 === 3) {
        card.style.backgroundColor = 'rgba(231, 90, 135, 1)';
      }
    });
  }, [cities]);

  const handleCityChange = async (e) => {
    const { value } = e.target;
    setSelectedCity(value);

    // Check if the input matches any existing city from the API data
    const matchingCity = Object.keys(weatherData).find(
      (city) => city.toLowerCase() === value.toLowerCase(),
    );

    // If the input matches an existing city, set filteredCities to contain only that matching city
    if (matchingCity) {
      setFilteredCities([matchingCity]);
    } else {
      // Filter the cities based on the input value
      const filtered = cities.filter((city) => city.toLowerCase().includes(value.toLowerCase()));

      // Update filteredCities with the filtered list
      setFilteredCities(filtered);
    }
  };

  return (
    <div className="container">
    <p className="title">Get live weather data by City name</p>
    <input
      type="text"
      placeholder="Enter city name"
      value={selectedCity}
      onChange={handleCityChange}
    />
    <ul className="cards">
      {filteredCities.map((city) => (
        <li key={city} className="card">
          <Link to={`/details/${city}`}>
            {weatherData[city] ? (
              <>
                <FontAwesomeIcon icon={faArrowCircleRight} className="angle-arrow" />
                <p>
                  <strong>City:</strong>
                  {' '}
                  {city}
                </p>
                <p>
                  <strong>Country:</strong>
                  {' '}
                  {weatherData[city].location.country}
                </p>
                <p>
                  <strong>Lat/Lon:</strong>
                  {' '}
                  {weatherData[city].location.lat}
                  /
                  {weatherData[city].location.lon}
                </p>
                <p>
                  <strong>tz_id:</strong>
                  {' '}
                  {weatherData[city].location.tz_id}
                </p>
              </>
            ) : (
              <div className="loading-spinner" />
            )}
          </Link>
        </li>
      ))}
    </ul>
  </div>
  );
}

export default Home;
