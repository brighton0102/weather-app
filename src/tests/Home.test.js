import React from 'react';
import { render } from '@testing-library/react';
import Home from '../components/Home';

test('Home component renders correctly', () => {
  const { container } = render(<Home />);
  expect(container).toMatchSnapshot();
});
