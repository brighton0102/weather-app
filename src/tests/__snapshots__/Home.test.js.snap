// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home component renders correctly 1`] = `
<div>
  <div
    class="container"
  >
    <p
      class="title"
    >
      Get live weather data by City name
    </p>
    <input
      placeholder="Enter city name"
      type="text"
      value=""
    />
    <ul
      class="cards"
    >
      <li
        class="card"
        style="background-color: rgb(255, 221, 170);"
      >
        <div
          class="loading-spinner"
        />
      </li>
      <li
        class="card"
      >
        <div
          class="loading-spinner"
        />
      </li>
      <li
        class="card"
      >
        <div
          class="loading-spinner"
        />
      </li>
      <li
        class="card"
        style="background-color: rgb(255, 221, 170);"
      >
        <div
          class="loading-spinner"
        />
      </li>
      <li
        class="card"
        style="background-color: rgb(255, 221, 170);"
      >
        <div
          class="loading-spinner"
        />
      </li>
      <li
        class="card"
      >
        <div
          class="loading-spinner"
        />
      </li>
      <li
        class="card"
      >
        <div
          class="loading-spinner"
        />
      </li>
      <li
        class="card"
        style="background-color: rgb(255, 221, 170);"
      >
        <div
          class="loading-spinner"
        />
      </li>
      <li
        class="card"
        style="background-color: rgb(255, 221, 170);"
      >
        <div
          class="loading-spinner"
        />
      </li>
      <li
        class="card"
      >
        <div
          class="loading-spinner"
        />
      </li>
      <li
        class="card"
      >
        <div
          class="loading-spinner"
        />
      </li>
      <li
        class="card"
        style="background-color: rgb(255, 221, 170);"
      >
        <div
          class="loading-spinner"
        />
      </li>
      <li
        class="card"
        style="background-color: rgb(255, 221, 170);"
      >
        <div
          class="loading-spinner"
        />
      </li>
      <li
        class="card"
      >
        <div
          class="loading-spinner"
        />
      </li>
      <li
        class="card"
      >
        <div
          class="loading-spinner"
        />
      </li>
      <li
        class="card"
        style="background-color: rgb(255, 221, 170);"
      >
        <div
          class="loading-spinner"
        />
      </li>
      <li
        class="card"
        style="background-color: rgb(255, 221, 170);"
      >
        <div
          class="loading-spinner"
        />
      </li>
      <li
        class="card"
      >
        <div
          class="loading-spinner"
        />
      </li>
      <li
        class="card"
      >
        <div
          class="loading-spinner"
        />
      </li>
      <li
        class="card"
        style="background-color: rgb(255, 221, 170);"
      >
        <div
          class="loading-spinner"
        />
      </li>
      <li
        class="card"
        style="background-color: rgb(255, 221, 170);"
      >
        <div
          class="loading-spinner"
        />
      </li>
      <li
        class="card"
      >
        <div
          class="loading-spinner"
        />
      </li>
      <li
        class="card"
      >
        <div
          class="loading-spinner"
        />
      </li>
      <li
        class="card"
        style="background-color: rgb(255, 221, 170);"
      >
        <div
          class="loading-spinner"
        />
      </li>
      <li
        class="card"
        style="background-color: rgb(255, 221, 170);"
      >
        <div
          class="loading-spinner"
        />
      </li>
      <li
        class="card"
      >
        <div
          class="loading-spinner"
        />
      </li>
      <li
        class="card"
      >
        <div
          class="loading-spinner"
        />
      </li>
      <li
        class="card"
        style="background-color: rgb(255, 221, 170);"
      >
        <div
          class="loading-spinner"
        />
      </li>
      <li
        class="card"
        style="background-color: rgb(255, 221, 170);"
      >
        <div
          class="loading-spinner"
        />
      </li>
      <li
        class="card"
      >
        <div
          class="loading-spinner"
        />
      </li>
      <li
        class="card"
      >
        <div
          class="loading-spinner"
        />
      </li>
      <li
        class="card"
        style="background-color: rgb(255, 221, 170);"
      >
        <div
          class="loading-spinner"
        />
      </li>
    </ul>
  </div>
</div>
`;
